community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/pipes/CachingExpandInto.scala

```diff
@@ -0,0 +1,147 @@
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.cypher.internal.compiler.v3_0.pipes
+
+import java.util.concurrent.ThreadLocalRandom
+
+import org.neo4j.cypher.internal.compiler.v3_0.ExecutionContext
+import org.neo4j.cypher.internal.compiler.v3_0.spi.QueryContext
+import org.neo4j.cypher.internal.frontend.v3_0.{InternalException, SemanticDirection}
+import org.neo4j.graphdb.{Relationship, Node}
+import org.neo4j.helpers.collection.PrefetchingIterator
+
+import scala.collection.mutable.ArrayBuffer
+import scala.collection.JavaConverters._
+import scala.collection.mutable
+
+/**
+ * Used by pipes that needs to expand between two known nodes.
+ *
+ * Given a pattern (a)-->(b) it will do the following:
+ * - if both a and b are dense nodes, start from the one with the lesser degree
+ * - if just one of the nodes is dense, start from the non-dense node
+ * - if both are non-dense, randomly pick one or the other
+ * - cache all found connecting relationships.
+ *
+ */
+trait CachingExpandInto {
+
+  /**
+   * Finds all relationships connecting fromNode and toNode.
+   */
+  protected def findRelationships(query: QueryContext, fromNode: Node, toNode: Node,
+                                relCache: RelationshipsCache, dir: SemanticDirection, relTypes: => Option[Seq[Int]]): Iterator[Relationship] = {
+
+    val fromNodeIsDense = query.nodeIsDense(fromNode.getId)
+    val toNodeIsDense = query.nodeIsDense(toNode.getId)
+
+    //if both nodes are dense, start from the one with the lesser degree
+    if (fromNodeIsDense && toNodeIsDense) {
+      //check degree and iterate from the node with smaller degree
+      val fromDegree = getDegree(fromNode, relTypes, dir, query)
+      if (fromDegree == 0) {
+        return Iterator.empty
+      }
+
+      val toDegree = getDegree(toNode, relTypes, dir.reversed, query)
+      if (toDegree == 0) {
+        return Iterator.empty
+      }
+
+      relIterator(query, fromNode, toNode, fromDegree < toDegree, relTypes, relCache, dir)
+    }
+    // iterate from a non-dense node
+    else if (toNodeIsDense)
+      relIterator(query, fromNode, toNode, preserveDirection = true, relTypes, relCache, dir)
+    else if (fromNodeIsDense)
+      relIterator(query, fromNode, toNode, preserveDirection = false, relTypes, relCache, dir)
+    //both nodes are non-dense, choose a random starting point
+    else
+      relIterator(query, fromNode, toNode, ThreadLocalRandom.current().nextBoolean(), relTypes, relCache, dir)
+  }
+
+  private def relIterator(query: QueryContext, fromNode: Node,  toNode: Node, preserveDirection: Boolean,
+                          relTypes: Option[Seq[Int]], relCache: RelationshipsCache, dir: SemanticDirection) = {
+    val (start, localDirection, end) = if(preserveDirection) (fromNode, dir, toNode) else (toNode, dir.reversed, fromNode)
```

Comment:
The same result yes, but if a is `a` has billions of relationships and `b` only has a couple of them - it is a lot faster to scan through b's incoming relationships looking for `a` on the other side than to scan through a's outgoing relationships looking for `b`.